File: drills/elixir/solutions/nc68_longest_palindrome__brute_force.exs
defmodule Solution do
  # Every start with every length. O(n^3) once the palindrome check is counted
  # -- the definition, and what centre expansion is an optimisation of.
  def longest_palindrome(""), do: ""

  def longest_palindrome(s) do
    n = String.length(s)

    for start <- 0..(n - 1)//1, length <- 1..(n - start)//1, reduce: "" do
      best ->
        candidate = String.slice(s, start, length)

        if String.length(candidate) > String.length(best) and
             candidate == String.reverse(candidate),
           do: candidate,
           else: best
    end
  end
end

File: drills/ts/solutions/nc68_longest_palindrome__brute_force.ts
export function longestPalindrome(s: string): string {
  // Every start with every length. O(n^3) once the palindrome check is counted
  // -- the definition, and what centre expansion is an optimisation of.
  let best = "";
  for (let start = 0; start < s.length; start++) {
    for (let end = start + 1; end <= s.length; end++) {
      const candidate = s.slice(start, end);
      if (candidate.length > best.length && candidate === [...candidate].reverse().join("")) {
        best = candidate;
      }
    }
  }
  return best;
}
